docs(state) add select and stateful tests, add docs, update readme (#82)
BioPhoton/rx-angular: libs/state/docs/snippets/deriving-simple-state.md

# Deriving simple state

> As the name implies this operator is useful when you process an Observable which maintains state.

Author: ```suggestion
As the name implies this operator is useful when you process an Observable which maintains state.
```

# Deriving simple state

> As the name implies this operator is useful when you process an Observable which maintains state.

Reviewer: What operator are you talking about?

> If an Observable maintain state there are repetitive and custom things.
> 
> The following things need to get ensured always:
> 
> - emit only distinct values (state changes)
> - filter out undefined values (lazy state)
> - share and replay the applied operations for multiple subscribers (display the state derivation)
> 
> The following things are always custom:
> 
> - get a subset of state derivation
> - compose in other Observables or change the Observables behaviour
> 
> If we take a simple state derivation and select the number of items in a list the above looks like this:

Reviewer: ```suggestion
Maintaining state as an `Observable` source comes with a handful of repetitive as well as situational tasks.

You will always (aka repetitive) want to ensure that:
- only distinct state changes are emitted
- only defined values are emitted (filter out undefined, which ensures lazy state)
- share and replay custom operations for multiple subscribers (saves performance)

You will sometimes (aka situational) need:
- a subset of the state (derivations)
- compose the state with other Observables or change the Observables behaviour

If we take a simple state derivation and select the number of items in a list the above looks like this:
```

```typescript
import { Observable } from 'rxjs';
import { distinctUntilChanged, map, filter, shareReplay } from 'rxjs/operators';

const state$: Observable<{ name: string; items: string[] }>;
const derivation$ = state$.pipe(
  // distinct same base-state objects (e.g. a emission of default switch cases, incorrect mutable handling of data) @TODO evaluate benefits vs. overhead
  distinctUntilChanged(),
  // CUSTOM LOGIC HERE
  map(state => state.list.length),
  // Filter out undefined to have lazy state and the ability to delete state slices over time
  filter(v => v !== undefined),
  // Distinct same values derived from the state
  distinctUntilChanged(),
  // Reuse custom operations result for multiple subscribers and reemit the last calculated value.
  shareReplay({ bufferSize: 1, refCount: true })
);
```

> By using the `stateful` operator we have the ability
> to place custom logic into the above process, and the repetitive operators get abstracted away.

Reviewer: ```suggestion
Using the `stateful` operator gives you the advantage to insert custom logic to derive state without having to think about sharing or replaying. It will also apply `distinctUntilChanged` by default. But you can provide custom logic for distinct values aswell.
```
